sw/device/lib/dif/dif_plic.h

```diff
@@ -0,0 +1,248 @@
+  dif_plic_target_t target;   /**< PLIC target */
+  dif_irq_data_t claimed_irq; /**< PLIC IRQ data */
+} dif_plic_t;
+
+/**
+ * Initialise an instance of PLIC
+ *
+ * Initialise a PLIC instance. Information that must be retained, and is
+ * required to program PLIC, shall be stored in @p arg3. PLIC supports external
+ * interrupt delivery to multiple targets, which are requested in @p arg2.
+ *
+ * Note that only target specific registers are initialised, the PLIC target
+ * common registers must be initialised by the caller.
+ * @param base_addr Base address of an instance of the PLIC IP block
+ * @param target Target that is interrupted by PLIC.
+ * @param plic PLIC state data
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_plic_init(uintptr_t base_addr, dif_plic_target_t target,
+                   dif_plic_t *plic);
+
+/**
+ * Enable/disable handling of IRQ source in PLIC.
+ *
+ * Enable/disable handling of IRQ source in PLIC. This operation does not affect
+ * the IRQ generation in the peripherals, which must be configured in a
+ * corresponding peripheral itself.
+ * @param plic PLIC state data
+ * @param enable Enable/disable the IRQ handling
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_plic_irq_ie_control(const dif_plic_t *plic, dif_plic_irq_src_id_t irq,
+                             dif_plic_on_off_t enable);
+
+/**
+ * Set IRQ trigger type (edge/level)
+ *
+ * Set the behaviour of the Interrupt Gateway for a particular IRQ to be edge
+ * or level triggered.
+ * @param plic PLIC state data
+ * @param irq Interrupt source ID
+ * @param enable Enable for the edge triggered, disable for level triggered IRQs
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_plic_irq_le_control(const dif_plic_t *plic, dif_plic_irq_src_id_t irq,
+                             dif_plic_on_off_t enable);
+
+/**
+ * Set IRQ source priority
+ *
+ * Set the IRQ source priority (0-3). In order for PLIC to set a CC register
+ * and assert the external interrupt line to the target (Ibex, ...), the
+ * priority of the IRQ source must be higher than the threshold for that source.
+ * @param plic PLIC state data
+ * @param irq Interrupt source ID
+ * @param priority Priority to be set
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_plic_irq_priority_set(const dif_plic_t *plic,
+                               dif_plic_irq_src_id_t irq, uint32_t priority);
+
+/**
+ * Get the IRQ interrupt pending status
+ *
+ * Get the status of the specific PLIC Interrupt Pending register IRQ bit.
+ * @param plic PLIC state data
+ * @param irq Interrupt source ID
+ * @param set Flag indicating whether the IRQ pending bit is set in PLIC
+ * @return true if the function was successful, false otherwise
+ */
+bool dif_plic_irq_ip_status(const dif_plic_t *plic, dif_plic_irq_src_id_t irq,
+                            bool *set);
```

Comment:
I am... extremely skeptical of this out parameter. This indicates to me that we should be putting better thought into our return types, so that you could return an enum like
```
enum dif_plic_irq_ip_status_err {
  kDifPlicIrqIpStatusOkEnabled,
  kDifPlicIrqIpStatusOkDisabled,
  kDifPlicIrqIpStatusErr...,
}
```
(The fact that C enums are not scoped is extremely obnoxious.)

Reply:
Agreed.